packages/field-base/src/label-controller.js

```diff
@@ -0,0 +1,208 @@
+/**
+ * @license
+ * Copyright (c) 2021 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { SlotController } from '@vaadin/component-base/src/slot-controller.js';
+
+/**
+ * A controller to manage the label element.
+ */
+export class LabelController extends SlotController {
+  constructor(host, labelId) {
+    super(
+      host,
+      'label',
+      () => document.createElement('label'),
+      (host, node) => {
+        this.label = host.label;
+
+        this.__savedLabelId = labelId;
+
+        // Set ID attribute or use the existing one.
+        this.__updateLabelId(node);
+
+        // Set text content for the default label.
+        this.__applyDefaultLabel(this.label, node);
+
+        this.__observeLabel(node);
+      }
+    );
+  }
+
+  hostConnected() {
+    super.hostConnected();
+
+    const labelNode = this.getSlotChild();
+    if (labelNode !== this.defaultNode) {
+      this.initCustomNode(labelNode);
+    }
+  }
+
+  /**
+   * Override to initialize the newly added custom label.
+   *
+   * @param {Node} labelNode
+   * @protected
+   * @override
+   */
+  initCustomNode(labelNode) {
+    this.__applyCustomLabel(labelNode);
+
+    this.__observeLabel(labelNode);
+  }
+
+  /**
+   * Override to cleanup label node when it's removed.
+   *
+   * @param {Node} _node
+   * @protected
+   * @override
+   */
+  teardownNode(_node) {
+    if (this.__labelObserver) {
+      this.__labelObserver.disconnect();
+    }
+
+    this.__applyDefaultLabel(this.label, this.defaultNode);
+  }
+
+  /**
+   * Set label based on corresponding host property.
+   * @param {string} label
+   */
+  setLabel(label) {
+    this.label = label;
+
+    if (this.defaultNode && this.node === this.defaultNode) {
+      this.__applyDefaultLabel(label, this.defaultNode);
+    }
+  }
+
+  /**
+   * Set callback to be called when label changes.
+   * @param {function()} callback
+   */
+  setLabelChangedCallback(callback) {
+    this.labelChangedCallback = callback;
+  }
+
+  /**
+   * @param {HTMLElement} labelNode
+   * @private
+   */
+  __applyCustomLabel(labelNode) {
+    this.__updateLabelId(labelNode);
+    const hasLabel = this.__hasLabel(labelNode);
+    this.__toggleHasLabel(hasLabel);
+  }
+
+  /**
+   * @param {HTMLElement} labelNode
+   * @return {boolean}
+   * @private
+   */
+  __hasLabel(labelNode) {
+    return Boolean(labelNode && (labelNode.children.length > 0 || this.__isNotEmpty(labelNode.textContent)));
+  }
+
+  /**
+   * @param {string} label
+   * @private
+   */
+  __isNotEmpty(label) {
+    return Boolean(label && label.trim() !== '');
+  }
+
+  /**
+   * @param {string} label
+   * @param {HTMLElement} labelNode
+   * @private
+   */
+  __applyDefaultLabel(label, labelNode) {
```

Comment:
It looks like this method has something of the `__applyCustomLabel` logic. To me, it would be better if this method was entirely about the default node.

Reply:
Thanks, fixed.